File: sed_repos/java/SED7/src/main/java/ic/doc/CachedForecaster.java
package ic.doc;

import java.util.LinkedHashMap;
import java.util.Map;

public class CachedForecaster implements ForecastGenerator {

  private final ForecastGenerator forecaster;
  private final Map<String, WeatherForecast> cache;

  public CachedForecaster(ForecastGenerator forecaster, int size) {
    this.forecaster = forecaster;

    this.cache = new LinkedHashMap<>() {
      protected boolean removeEldestEntry(Map.Entry<String, WeatherForecast> eldest) {
        return size() > size;
      }
    };
  }


  @Override
  public WeatherForecast forecastFor(WeatherRegion region, WeatherDay day) {
    String searchString = region + " : " + day.toString();
    WeatherForecast forecast = cache.get(searchString);
    if (forecast == null) {
      forecast = forecaster.forecastFor(region, day);
      cache.put(searchString, forecast);
    }

    if (forecast == null) {
      throw new RuntimeException("Could not find a forecast");
    }

    return forecast;
  }
}

File: sed_repos/python/SED7/src/cached_forecaster.py
from src.forecast_generator import ForecastGenerator


class CachedForecaster(ForecastGenerator):
    def __init__(self, forecaster, size):
        self.forecaster = forecaster
        self.cache = {}
        self.size = size

    def forecast_for(self, region, day):
        key = f"{region} : {day}"
        if key not in self.cache:
            forecast = self.forecaster.forecast_for(region, day)
            if forecast is None:
                raise RuntimeError("Could not find a forecast")
            if len(self.cache) >= self.size:
                self.cache.pop(next(iter(self.cache)))
            self.cache[key] = forecast
        return self.cache[key]
